### PR TITLE
Call openReport when parent report is mssiing

### PR DESCRIPTION
**Pull request by shahinyan11**
### Fixed Issues

$ https://github.com/Expensify/App/issues/34346
PROPOSAL: https://github.com/Expensify/App/issues/34346#issuecomment-1934568395

### Tests

1. Create a new account, lets call it Account A
2. Create another new account, account B
3. Send an IOU from account A to account B and go to IOU details page and copy the URL
4. Sign in with account B
5. Navigate to the copied URL on step 3
6. Verify that the parent report link is exists in header

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as in the Tests section.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/46921547/5681aa71-ac2f-4942-bd55-702c8dc6d494

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by shahinyan11** [comment]
@s77rt The changes are not platform specified and those are functional and not affect to UI.  Can I add only one video ?

File: src/pages/home/ReportScreen.js [hunk 1]
```diff
@@ -354,7 +355,13 @@ function ReportScreen({
         // It possible that we may not have the report object yet in Onyx yet e.g. we navigated to a URL for an accessible report that
         // is not stored locally yet. If report.reportID exists, then the report has been stored locally and nothing more needs to be done.
         // If it doesn't exist, then we fetch the report from the API.
-        if (report.reportID && report.reportID === getReportID(route) && !isLoadingInitialReportActions) {
```
Comment (s77rt) [inline src/pages/home/ReportScreen.js:357]: We shouldn't remove this condition. just add the new logic without breaking the current one

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/home/ReportScreen.js (hunk 1 above)
Comment (shahinyan11) [reply src/pages/home/ReportScreen.js]: sorry, I mixed up the `if` blocks

File: src/pages/home/ReportScreen.js [hunk 2]
```diff
@@ -348,6 +349,12 @@ function ReportScreen({
         // Report ID will be empty when the reports collection is empty.
         // This could happen when we are loading the collection for the first time after logging in.
         if (!ReportUtils.isValidReportIDFromPath(reportIDFromPath)) {
+            const parentReport = ReportUtils.getReport(report);
+
+            if (ReportUtils.isThread(report) && isEmptyObject(parentReport)) {
+                Report.openReport(report.parentReportID);
+            }
+
```
Comment (s77rt) [inline src/pages/home/ReportScreen.js:357]: This still does not look right. Why are we opening the parent report if the report id from path is not valid?

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/home/ReportScreen.js (hunk 2 above)
Comment (shahinyan11) [reply src/pages/home/ReportScreen.js:357]:
```
if (report.reportID && report.reportID === getReportID(route) && !isLoadingInitialReportActions) {
    const parentReport = ReportUtils.getReport(report)

    if (ReportUtils.isThread(report) && isEmptyObject(parentReport)) {
        Report.openReport(report.parentReportID)
    }
    
    return;
}
```

Comment (shahinyan11) [reply src/pages/home/ReportScreen.js:357]: @s77rt  Do you approve of the above options?

File: src/pages/home/ReportScreen.js [hunk 3]
```diff
@@ -355,11 +356,17 @@ function ReportScreen({
         // is not stored locally yet. If report.reportID exists, then the report has been stored locally and nothing more needs to be done.
         // If it doesn't exist, then we fetch the report from the API.
         if (report.reportID && report.reportID === getReportID(route) && !isLoadingInitialReportActions) {
+            const parentReport = ReportUtils.getParentReport(report);
+
+            if (ReportUtils.isThread(report) && isEmptyObject(parentReport)) {
+                Report.openReport(report.parentReportID);
+            }
+
```
Comment (s77rt) [inline src/pages/home/ReportScreen.js:364]: Can you explain the logic behind this? i.e. why it's in that if block

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/home/ReportScreen.js (hunk 3 above)
Comment (shahinyan11) [reply src/pages/home/ReportScreen.js:364]: @s77rt I thought it worth to call new code when the report exists

**Comment by s77rt** [comment]
Looks we have more bugs to cover
1. Initially I'm getting the not found view
2. After reload, the `[Deleted message]` title can be briefly seen
3. Money request info is missing and had to reload another time to get the correct view

https://github.com/Expensify/App/assets/16493223/db09a1a1-3bd6-456b-a5f3-976c4856b47a

**Comment by shahinyan11** [comment]
Ok . I will work on it

**Comment by s77rt** [comment]
> I think the ReportScreen component should be updated and use withReportOrNotFound

I don't see how would that fix the bug?

> It does not reproducible in last main

Can you please merge main to this PR

> The root cause is the memoization of component

Memorization being the root cause would make sense (didn't check in depth) but we should not render every report action just because the parent action is different (especially since many report actions have no relation with the report's parent action)

**Comment by shahinyan11** [comment]
> > I think the ReportScreen component should be updated and use withReportOrNotFound
> 
> I don't see how would that fix the bug?

Doesn't it requests the report when it doesn't find it in Onyx?

> > The root cause is the memoization of component
> 
> Memorization being the root cause would make sense (didn't check in depth) but we should not render every report action just because the parent action is different (especially since many report actions have no relation with the report's parent action)

Then we can use below comparison
```
ReportActionsUtils.isTransactionThread(ReportActionsUtils.getParentReportAction(prevProps.report)) !== ReportActionsUtils.isTransactionThread(ReportActionsUtils.getParentReportAction(nextProps.report))
```

**Comment by s77rt** [comment]
Seeing that we have more bugs to deal with than expected I think it's best to rethink how we can fix those bugs on the drawing board i.e. the github issue

**Comment by s77rt** [comment]
@shahinyan11 I have replied on the linked issue. Can you please close this PR and post new proposals on the issue covering all the found bugs

**Comment by shahinyan11** [comment]
> @shahinyan11 I have replied on the linked issue. Can you please close this PR and post new proposals on the issue covering all the found bugs

Are you able to reproduce `[Deleted message]` after las updates ?

**Comment by s77rt** [comment]
@shahinyan11 Just retested and I didn't get the `[Deleted Message|` bug but the rest are still reproducible


https://github.com/Expensify/App/assets/16493223/b3a46c73-e104-4694-832d-d2e53bfbf820

**Comment by s77rt** [comment]
@shahinyan11 Usually we treat each bug separately and in parallel but here we can't fix the wrong header bug without fixing the not found view bug first. They do not have same root cause so I guess the proposals should outline each bug independently. We can open another issue for the not found view bug but since they have the same reproduction steps it may be good to consolidate them (as long as they are not too complex).

**Comment by s77rt** [comment]
@shahinyan11 Can you please close this PR

**Pull request closed without merging**